### PR TITLE
[tune] Add xgboost_ray integration

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Adds reporting utilities to the xgboost callback for interoperability with `xgboost_ray`

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/integration/xgboost.py [hunk 1]
```diff
@@ -3,6 +3,20 @@
 
 import os
 
+try:
+    from xgboost_ray.session import is_xgboost_ray_actor, get_actor_rank, \
+        put_queue
+except ImportError:
+
```
Comment (richardliaw) [inline python/ray/tune/integration/xgboost.py:10]: Can you add some comments here about what you are doing (just for external users to read)

**Comment by richardliaw** [comment]
Looks good otherwise, would like @amogkam to take a look though.

**Comment by amogkam** [comment]
@krfricke It might be better to not add xgboost_ray logic in here and to keep them separate. One thing we can do is modify these callbacks like so:
```
class TuneReportCallback(TuneCallback):
    def __init__(self,
                 metrics: Union[None, str, List[str], Dict[str, str]] = None):
        if isinstance(metrics, str):
            metrics = [metrics]
        self._metrics = metrics

    def _get_result(self, env):
        result_dict = dict(env.evaluation_result_list)
        if not self._metrics:
            report_dict = result_dict
        else:
            report_dict = {}
            for key in self._metrics:
                if isinstance(self._metrics, dict):
                    metric = self._metrics[key]
                else:
                    metric = key
                report_dict[key] = result_dict[metric]
        return report_dict
 
  def __call__(self, env):
    report_dict = self._get_result(env)
    tune.report(**report_dict)
```
(and something similar for the CheckpointCallback). And then in xgboost_ray, we can subclass these and just override the __call__ method. Let me know what you think.

**Comment by richardliaw** [comment]
@amogkam , good discussion - can you provide some reasoning here about why you would prefer your suggested approach?

**Comment by amogkam** [comment]
@richardliaw the main thing is, like you said, it gets confusing adding logic from upstream libraries into Ray and it sort of adds a 2 way dependency: xgboost_ray depends on Ray functionality, but now Ray depends on xgboost_ray functionality like `put_queue`. It's also a slippery slope in terms of what lengths we would go to in order to support external libraries. Imagine if xgboost_ray was not created by us but by a 3rd party. In that case I don't think we would accept this PR, but what we would do is slightly modify our implementation to making subclassing on their side easier.

**Comment by amogkam** [comment]
In a broader sense, I don't think this new functionality makes sense for a general purpose Xgboost+Tune callback. It's specific to the xgboost_ray implementation so I feel like it should belong in that library.

**Comment by krfricke** [comment]
Strictly speaking we're not introducing a dependency to xgboost_ray as we handle import errors. I agree however that this seems like an exception targeted to better interoperability with our own contributed libraries.

The reason why I introduced this was that users would be able to use the normal Tune-provided callbacks without worrying about interoperability. It seems confusing if they have to use different callbacks when they use xgboost_ray instead of regular xgboost, and this seemed like a common source for errors.

I guess we could avoid this problem by introducing the specific callbacks as suggested by you, Amog (i.e. keep ray "clean" and add a new callback in xgboost_ray) and looping through the callbacks in xgboost_ray, replacing regular tune callbacks with xgboost_ray specific callbacks when appropriate. What do you think?

If there are no objections I will adjust both PRs accordingly.

**Comment by amogkam** [comment]
Wait currently users don’t even have to add the tune callbacks for xgboost_ray right? We detect if they are in a session and then automatically add the callback for them if so.

**Comment by krfricke** [comment]
We do that for the reporting callback but not for the checkpointing callback. And if they pass one of those manually we should still replace it, or not?

**Comment by amogkam** [comment]
Here's what I propose for now: 
- Modify these callbacks to make them easier to extend
- Subclass these callbacks in xgboost_ray
- For now the user can just manually use the subclassed callbacks from xgboost_ray. If no callbacks are provided but we are in a tune session, we will still automatically add the reporting callback.

I agree with you Kai that having separate callbacks is confusing and can potentially lead to errors. But I think for now it's fine since it doesn't seem like anyone is using the xgboost_ray+tune integration yet. But we should discuss this further. Some options include:
- Kai's approach of replacing the standard xgboost+tune callbacks with the xgboost_ray+tune callbacks
- Just having the users use the xgboost_ray callbacks directly and raising an error early if they use the standard xgboost+tune callbacks.
- Somehow handling all of this automatically so the user doesn't even have to use callbacks if they are in a tune session. Perhaps if there's a way to inspect the signature of the caller function to see if `checkpoint_dir` is being accepted.

Also some other questions to consider:
- Can we use the same checkpoint for failure handling and for tune?
- Should the checkpoint_freq in RayParams also impact Tune checkpoint frequency? 

@krfricke @richardliaw what do you guys think?

**Comment by richardliaw** [comment]
> Here's what I propose for now:
> 
> * Modify these callbacks to make them easier to extend
> * Subclass these callbacks in xgboost_ray
> * For now the user can just manually use the subclassed callbacks from xgboost_ray. If no callbacks are provided but we are in a tune session, we will still automatically add the reporting callback.

This sounds fine! 

> I agree with you Kai that having separate callbacks is confusing and can potentially lead to errors. 

I also think it is confusing, but this is not a [type 1 decision](https://www.businessinsider.com/jeff-bezos-on-type-1-and-type-2-decisions-2016-4). So I'm fine with the Amog's proposal which seems to be a bit less complex (no logic referencing xgboost-ray in Tune).

> But I think for now it's fine since it doesn't seem like anyone is using the xgboost_ray+tune integration yet. But we should discuss this further.

I think this line of reasoning is a bit of a fallible -- it's saying that it's ok to cut corners on usability **because** we don't have users yet. If you don't plan to have users/know when you're going to have users, you basically receive a blank check to do anything. On the other hand, we have already identified our users and are going to have real, daily users soon right? 

Instead, I think it's ok to cut this corner because we can provide a simple workaround (like raising an error early as you suggested in your followup points).

**Comment by krfricke** [comment]
Great. Since the callbacks will be very close to the original callbacks, I'll replace them and raise a warning, so the code works out of the box. I think this is preferable because early in the development cycle you might want to call a training function without Tune, before wrapping it with Tune for HPO - this way you can just keep the very same code.

Detecting `checkpoint_dir` in the trainable seems like a good idea. 

We probably won't be able to use the same checkpoints for experiment progress and Tune, but we can investigate this further. Let's move that discussion to the xgboost_ray repo though.

I'll update the PRs today.

File: python/ray/tune/integration/xgboost.py [hunk 2]
```diff
@@ -81,15 +86,24 @@ class _TuneCheckpointCallback(TuneCallback):
...
 
     """
 
-    def __init__(self, filename: str = "checkpoint"):
+    def __init__(self, filename: str = "checkpoint", frequency: int = 5):
         self._filename = filename
+        self._frequency = frequency
 
-    def __call__(self, env):
+    @staticmethod
```
Comment (amogkam) [inline python/ray/tune/integration/xgboost.py:98]: why does this have to be a static method?

**Review by amogkam — APPROVED** [review]
LGTM, thanks @krfricke! Just left one question about static method.

File: python/ray/tune/integration/xgboost.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/tune/integration/xgboost.py:98]: Hmm, looks like it might not need to be, but I'll merge first as it's not too consequential.

Comment (krfricke) [reply python/ray/tune/integration/xgboost.py:98]: The idea here is that in xgboost_ray we'll send this callback to the driver via the queue. A non-static method will pickle `self` when it's being sent via the object store, while a static method shouldn't do so.

**Pull request merged**